source/automated-deployment.rst

```diff
@@ -0,0 +1,131 @@
+    collection stack installed on
+* BEAMLINE_SERVER
+    The main "compute" computer where jupyterhub will point to
+* BEAMLINE_USER
+    The beamline user name. `xf23id1`, `xf11id`, etc...
+* IPYTHON_CONFIGURATION:
+    The files that are used to configure the data collection environment.
+    These have thus far been a collection of files that are prefixed with
+    two numbers and are stored in ~/.ipython/profile_collection or a variant
+* IPYTHON_CONFIGURATION_LOCATION:
+    Proposed to be at `~/ipython_profile`. Or maybe `~/bluesky_configuration` is better
+* IPYTHON_CONFIGURATION_NAME:
+    Currently it is `collection` at most places. I think we should standardize
+    on this.
+* {{ BEAMLINE }}
+    Placeholder syntax for any beamline ID, i.e., LIX, CHX, XPD, CSX, etc...
+* {{ BEAMLINE }}_CONFIGURATION
+    Conda package that includes the configuration information for at least
+    metadatastore and filestore.  Optionally includes configuration information
+    for a local channelarchiver server.  Other configuration information is
+    welcome to be put in here.  BACKUP_LOCATION is something that should be
+    added into the this configuration package
+* COLLECTION
+    Conda metapackage that enumerates the conda packages that make up our
+    collection environment
+* COLLECTION_{{ BEAMLINE }}
+    Conda metapackages that are beamline specific and include any extra
+    packages that a beamline needs or wants installed in their collection
+    environment.
+* ANALYSIS
+    conda metapackage that enumerates the conda packages that make up our
+    analysis environment
+* ANALYSIS_{{ BEAMLINE }}
+    conda metapackage that is beamline specific and includes at least the
+    analysis metapackage and the {{ BEAMLINE }}_configuration conda package
+* OLOG
+    Logbook for everything {{ BEAMLINE }} related
+* UAT
+    User acceptance tests. These are the tests that, when successfully run,
+    allows us to declare victory that our software stack is still working as
+    intended on each beamline.  Note that these UAT's are the beamline's
+    responsibility to create and maintain as it is just simply not possible
+    for the DAMA group to know every way that each beamline is using the
+    software stack.
+
+Automated installation procedure for collection environment
+-----------------------------------------------------------
+0. Define some bash functions to make notifications easier... ::
+
+    notify_slack ()
+    {
+        curl -d "title=$1&body=$2&room=$3" http://penelope:5000/message
+    }
+
+    log_to_olog ()
+    {
+        # log the files that were changed to the logbook.
+        su - {{ BEAMLINE_USER }}
+        source activate collection
+        python -c "from pyOlog.cli.ipy import olog; import os; olog(msg=$1, logbooks='DAMA Software')"
+
+    }
+1. Commit all python files that are in {{ IPYTHON_CONFIGURATION_LOCATION }}/profile_{{ IPYTHON_CONFIGURATION_NAME }} ::
```

Comment:
If we use 'cp' for backup, make it 'cp -rp'.


On startup we could notify about:
  - newer versions
  - untracked files or uncommited changes
  - packages installed that aren't in nsls2-<beamline>
?


Puppet runs every 30 minutes, not weekly.  To puppetify stuff, we'll
need a lightweight, idempotent script that checks some location for
what a beamline is supposed to have and ensures that's the case.

The sysadmins would typically stick info like that into hostgroups in
Foreman, but seems like we could use git tags here and avoid the need
for another tool.  Changing a beamline would involve coordinating
updating the 'production', or whatever, tag(s) with the beamline's
schedule, and then within ~30 minutes it's automatically taken care of.

The benefit of puppet over using something manually kicked off is
puppet will catch systems that were down during a manual run aswell as
new systems.

-matt